File: packages/core/src/action.js
import { BOARD_SIZE } from "./constants.js";
import { ACTION_SIZE, PASS_ACTION } from "./actionSpace.js";
import { ORIENTATIONS } from "./orientation.js";

export function encodeAction(move) {
  if (move.kind === "pass") return PASS_ACTION;
  return move.orientationGlobalId * BOARD_SIZE * BOARD_SIZE + move.y * BOARD_SIZE + move.x;
}

export function decodeAction(action, player) {
  if (action === PASS_ACTION) return { kind: "pass", player };
  if (action < 0 || action >= ACTION_SIZE) {
    throw new Error(`Action out of range: ${action}`);
  }
  const orientationGlobalId = Math.floor(action / (BOARD_SIZE * BOARD_SIZE));
  const position = action % (BOARD_SIZE * BOARD_SIZE);
  const orientation = ORIENTATIONS[orientationGlobalId];
  return {
    kind: "place",
    player,
    pieceId: orientation.pieceId,
    orientationGlobalId,
    x: position % BOARD_SIZE,
    y: Math.floor(position / BOARD_SIZE),
  };
}

export function legalMaskFromMoves(moves) {
  const mask = new Uint8Array(ACTION_SIZE);
  for (const move of moves) {
    mask[encodeAction(move)] = 1;
  }
  return mask;
}

File: training/native/blokus-selfplay/src/action.rs
//! Mirror of `packages/core/src/action.js` + `actionSpace.js`.
//!
//! Action encoding (must match the JS/Python action space exactly):
//!   place => orientationGlobalId * 196 + y * 14 + x
//!   pass  => PASS_ACTION (= 91 * 196 = 17836)
//!   ACTION_SIZE = 91 * 196 + 1 = 17837

use crate::constants::BOARD_SIZE;
use crate::orientation::{get_orientation, orientation_count};
use crate::rules::Move;

/// Number of board positions an orientation anchor can occupy (14 * 14 = 196).
pub fn num_board_positions() -> usize {
    BOARD_SIZE * BOARD_SIZE
}

/// 91 * 196 = 17836.
pub fn place_actions() -> usize {
    orientation_count() * num_board_positions()
}

/// The pass action id (= place_actions()).
pub fn pass_action() -> usize {
    place_actions()
}

/// Total action-space size (= place_actions() + 1).
pub fn action_size() -> usize {
    place_actions() + 1
}

/// Encode a move to its action id. Mirrors `encodeAction`.
pub fn encode_action(mv: &Move) -> usize {
    match mv {
        Move::Pass { .. } => pass_action(),
        Move::Place {
            orientation_global_id,
            x,
            y,
            ..
        } => orientation_global_id * BOARD_SIZE * BOARD_SIZE + y * BOARD_SIZE + x,
    }
}

/// Decode an action id (for `player`) back to a move. Mirrors `decodeAction`.
///
/// Panics on out-of-range action, like the JS implementation throws.
pub fn decode_action(action: usize, player: u8) -> Move {
    if action == pass_action() {
        return Move::Pass { player };
    }
    if action >= action_size() {
        panic!("Action out of range: {action}");
    }
    let positions = BOARD_SIZE * BOARD_SIZE;
    let orientation_global_id = action / positions;
    let position = action % positions;
    let orientation =
        get_orientation(orientation_global_id).expect("orientation id within range");
    Move::Place {
        player,
        piece_id: orientation.piece_id.clone(),
        orientation_global_id,
        x: position % BOARD_SIZE,
        y: position / BOARD_SIZE,
    }
}
